core/src/main/java/org/jruby/runtime/InterpretedIRBlockBody.java

```diff
@@ -93,31 +92,28 @@ public String getName() {
 
     @Override
     public boolean canCallDirect() {
-        return interpreterContext != null && interpreterContext.hasExplicitCallProtocol();
+        return fullInterpreterContext != null && fullInterpreterContext.hasExplicitCallProtocol();
     }
 
     @Override
     protected IRubyObject callDirect(ThreadContext context, Block block, IRubyObject[] args, Block blockArg) {
-        InterpreterContext ic = ensureInstrsReady(); // so we get debugging output
-        return Interpreter.INTERPRET_BLOCK(context, block, null, ic, args, block.getBinding().getMethod(), blockArg);
+        ensureInstrsReady(); // so we get debugging output
+        return Interpreter.INTERPRET_BLOCK(context, block, null, fullInterpreterContext, args, block.getBinding().getMethod(), blockArg);
     }
 
     @Override
     protected IRubyObject yieldDirect(ThreadContext context, Block block, IRubyObject[] args, IRubyObject self) {
-        InterpreterContext ic = ensureInstrsReady(); // so we get debugging output
-        return Interpreter.INTERPRET_BLOCK(context, block, self, ic, args, block.getBinding().getMethod(), Block.NULL_BLOCK);
+        ensureInstrsReady(); // so we get debugging output
+        return Interpreter.INTERPRET_BLOCK(context, block, self, fullInterpreterContext, args, block.getBinding().getMethod(), Block.NULL_BLOCK);
     }
 
     @Override
     protected IRubyObject commonYieldPath(ThreadContext context, Block block, Block.Type type, IRubyObject[] args, IRubyObject self, Block blockArg) {
         if (callCount >= 0) promoteToFullBuild(context);
 
-        InterpreterContext ic = ensureInstrsReady();
+        ensureInstrsReady();
```

Comment:
My only question is it wrong for 115 to stay as it is while still removing line 120 from before?  I see a field lookup after ensureInstrsReady() in new patch and I am trying to reason through concurrency aspects of that vs returning from ensureInstrsReady.  All I can think of it if somehow two ensureInstrsReady happen at the same time you will just get one. They both at least logically should be the same.

Reply:
You're right, and I've pushed a change that uses the result of calling `ensureInstrsReady`. Since there's no crossover on the two IC fields, this is now safe (and there's no reason to re-read `interpreterContext`).